File: 0001-0500/0124.Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxPathSum(self, root, isTop=True):
        """
        :type root: TreeNode
        :rtype: int
        """
        if root is None:
            return float('-inf')
        if isTop:
            self.result = root.val        
        left = self.maxPathSum(root.left, False)
        right = self.maxPathSum(root.right, False)
        rtn = max(left + root.val, right + root.val, root.val)
        self.result = max(self.result, left + right + root.val, rtn)
        if isTop:
            return self.result
        return rtn

File: 0001-0500/0124.Binary Tree Maximum Path Sum.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def max_path_sum(root)
    @res = root.val
    def dfs(root)
        if root
            l, r = dfs(root.left), dfs(root.right)
            l, r = [l, 0].max, [r, 0].max
            @res = [@res, l + r + root.val].max
            [l + root.val, r + root.val].max
        else
            -1.0 / 0.0
        end
    end
    dfs root
    @res
end
